### PR TITLE
Making code less SDSS dependent : part one

### PR DESCRIPTION
**Pull request by julianbautista**
In this PR, there are several things being implemented:

1) simplifying io.py with the usage of astropy.table.Table to store the metadata. This makes it easy to propagate metadata for SDSS or DESI or other types. There are some modes that I didn't update yet (I wonder if anyone uses them).

2) removed some warning messages in picca_deltas.py

3) removed the dependency from Delta in picca_deltas.py

4) improving the readability of the logs in picca_deltas.py (e.g, when it prints the var_lss, eta, fudge, etc)

The next step is to actually modify the picca.data.QSO and Forest classes to remove the explicit need of plate, mjd, fiber.

### CODE REVIEW

File: py/picca/io.py
Comment (Waelthus) [inline py/picca/io.py:1205]: why the "*1"? I don't think this is the same operation.

**Review by Waelthus — CHANGES_REQUESTED** [review]
Looks good in general, there's still a bunch of statements saying "Not implemented" in `io.py`, so those need to be fixed for sure. Also some changes might break compatibility with earlier code versions, so people would need to be informed before merging as previous analysis cannot be run anymore without chanegs of either catalogs or submission scripts.

**Comment by Waelthus** [comment]
@andreufont : I don't think we can really rely on travis for all the hiccups, there's just too many options people might use that are currently not covered in the test, and some will fail, e.g. running on desi mocks using a DRQ. I guess we should actually start working on improving the tests before the next big PR.

File: bin/picca_deltas.py
Comment (julianbautista) [reply bin/picca_deltas.py:114]: Yes. The Delta class is barely used for continuum fitting. 
The goal is to break the code into three main components (continuum fitting, correlations, fitters). 
Yes, there are no other uses of Delta in this part of the code.

File: bin/picca_deltas.py
Comment (julianbautista) [reply bin/picca_deltas.py:463]: yes, fixed.

File: py/picca/prep_del.py
Comment (julianbautista) [reply py/picca/prep_del.py:232]: yes, this removes a Deprecation Warning. The warning itself suggested the change!

File: py/picca/data.py
Comment (julianbautista) [reply py/picca/data.py:459]: This is in Forest, isn't it?

File: bin/picca_deltas.py
Comment (julianbautista) [reply bin/picca_deltas.py:497]: I'm not sure I understand the suggestion. best_obs, single_exp, spall are options for some of the read_from_* functions.

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py:157]: Yes. Could you send me a script where this is run on desi mocks? Or point me to someone that does that. I agree this is not covered by Travis and has to be added. 
I added 
`if 'desi' in mode and 'TARGETID' in catalog.colnames:`
which should cover the cases you mention.

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py:1205]: I want to make sure we are doing copies of the data. 
Also I didn't understand the operations with ivar (some weighted average).

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py:1118]: Ok fine. I will leave it as a comment and put back your glob for now.

File: bin/picca_deltas.py
Comment (julianbautista) [reply bin/picca_deltas.py:531]: Yes. It's still a lot of noisy code. I was planning to send a Table directly to forest.mask instead of these arrays.

**Comment by julianbautista** [comment]
@andreufont yes, sorry, I started doing one change and I did others in between, but they are quite minor (warnings).

I'm running DR16 tests but on S82 data and comparing with master. I'm also running DESI mini SV. 
I'm bugging other people to provide tests for mocks. 

In any case, with git is easy to come back to previous versions and track problems.

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py:519]: yes, that would actually be a good idea to keep everything more clean. Generating one table of those and then passing them to mask and add_dla and getting the right columns from there. This would also be more flexible in case one would like to later add additional stuff to the masking, e.g. BAL_RF or something...

File: py/picca/prep_del.py
Comment (Waelthus) [reply py/picca/prep_del.py:232]: ok, then it should be fine.

File: py/picca/data.py
Comment (Waelthus) [reply py/picca/data.py:459]: Oh, you're right. My bad. Yes, remove it from the initializer (which allows getting rid of it in `read_from_*`), but we need to set it manually somewhere else.

Comment (julianbautista) [reply py/picca/data.py:459]: Yes, it's in picca_deltas.py. I set them after read_data().

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py:485]: What I meant is that you removed the order from the initilizer of Forest and as an argument for all the `read_from_*` routines, which I guess makes things less messy. But I'd suggest to still set the order in io.read_data(...), e.g. just before returning using the same 4 lines you've used here to set it. The reason would be to keep the picca_deltas cleaner as this doesn't actually do anything with the forests yet.

File: py/picca/io.py
Comment (Waelthus) [reply py/picca/io.py:157]: about the script, please remind me when NERSC is back,

about the change to `if 'desi' in mode and 'TARGETID' in catalog.colnames` this would keep things in the cat on eBOSS names, which I guess would be removed in `read_from_desi`, that's why I suggested the mapping

File: py/picca/io.py
Comment (Waelthus) [reply py/picca/io.py:1212]: Ok, I think the sum was used here to do an average over all selected spectra (in case one has the same targetid in the sample mulitple times), that converts an [nspec, ndiag, npix] array to an [ndiag, npix] array so cannot be removed here. Also the *1 for a copy is not really good style ;-) why not use .copy() instead

All the ivar operations are to do inverse variance weighted coadding between different spectra with same THINGID/TARGETID within the same coadd-file. The actual coadd routine, currently only coadds the different arms I think.

All of those routines probaby need a bit of an update to be more clear...

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py]: Done

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py]: Done

File: bin/picca_deltas.py
Comment (iprafols) [inline bin/picca_deltas.py:444]: why are we removing the try/except block here? I think it is useful to inform the user if an error is occuring here

Comment (iprafols) [inline bin/picca_deltas.py:444]: hdul here should be renamed to hdu since it is now just a Header Data Unit, and not a Header Data Unit List

File: bin/picca_deltas.py
Comment (iprafols) [reply bin/picca_deltas.py:485]: I'm not sure what would be the best place for this. It is true that moving it to read_data(...) would keep picca_deltas cleaner, but adding it to read_data(...) feels weird in itself since the order is not actually read. I think this may be why Julian chose to move it here.

How about we add this as a method `set_order` of Forest? It would have the save 4 lines. I think this way we could have the benefits from all points of view

File: bin/picca_deltas.py
Comment (iprafols) [reply bin/picca_deltas.py:519]: You seem to agree on this (and I do as well), but are we thinking to add this in part 2?

File: bin/picca_deltas.py
Comment (iprafols) [inline bin/picca_deltas.py:764]: Why are you removing the list comprehension? How does this affect performance?

File: py/picca/data.py
Comment (iprafols) [inline py/picca/data.py:16]: If I recall properly, this was left as a function since it was used in multiple places. I understand removing it and making the computation explicitly in class Forest but, what happened to the other places this was used?

File: py/picca/data.py
Comment (iprafols) [reply py/picca/data.py:459]: This is addressed by the other question in picca_deltas

File: py/picca/io.py
Comment (iprafols) [inline py/picca/io.py:186]: instead of `keep_bal == False` the standard is to type `not keep_bal`

File: py/picca/io.py
Comment (iprafols) [inline py/picca/io.py:235]: `hdul` here should be renamed to `hdu`

File: py/picca/io.py
Comment (iprafols) [inline py/picca/io.py:583]: why not use the .format() ? I think it is a bit clearer than the f"  " notation

File: py/picca/io.py
Comment (iprafols) [inline py/picca/io.py:612]: I don't think so, but I would keep it for a bit longer. Maybe we can add a DeprecationWarning if this function is used

File: py/picca/io.py
Comment (iprafols) [inline py/picca/io.py:1202]: why are you not using copy here but using it on line 1057?

File: py/picca/data.py
Comment (iprafols) [reply py/picca/data.py:16]: Ok, I saw that you modified the other places as well

**Review by iprafols — CHANGES_REQUESTED** [review]

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py:1202]: We should probably use copy just to make sure. I'm adding it back

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py:583]: I disagree. How to solve this conflict ? :)

File: bin/picca_deltas.py
Comment (julianbautista) [reply bin/picca_deltas.py:444]: I removed because I was having the error message but it was not informative enough. 
Removing the try/except block I have more information on the error and the code stops anyway.

Comment (julianbautista) [reply bin/picca_deltas.py:444]: seriously...

File: py/picca/io.py
Comment (iprafols) [reply py/picca/io.py:583]: I'm not strong headed here. I guess we can leave this with the f notation, but maybe @Waelthus can break the tie

File: bin/picca_deltas.py
Comment (iprafols) [reply bin/picca_deltas.py:444]: hahahha, you can leave it as hdul if you want, but this was just to let you know my reasoning behind using these two names in the rest of the code

Comment (iprafols) [reply bin/picca_deltas.py:444]: Ok, I see

File: bin/picca_deltas.py
Comment (julianbautista) [reply bin/picca_deltas.py:764]: Because we don't create Delta objects anymore. We modify them. 
Performance here is definitely not a bottleneck.

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py:186]: ok

File: bin/picca_deltas.py
Comment (iprafols) [reply bin/picca_deltas.py:779]: makes sense

**Review by iprafols — COMMENTED** [review]
I approve this changes, but there are still unresolved questions that should be addressed before merging, particularly those signaled by Michael

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py]: Agreed, why rename all variables to clearer names if immediately after standardization we stop following the naming rules ;-)

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py:528]: Couldn't one even add it in here? Sounds sufficiently simple, I'll have a look how much of a change this would be

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py:489]: This does sound like a good compromise to me. Another option that just came to my mind would be to have the data dictionary actually be an object to allow having a way to put all the 
```
for (heal)pix in data:
        for forest in data[(heal)pix]:
```
constructions into a function (or iterator or something), maybe even having parameters that are dataset and not really spectrum dependent such as e.g. order in there. But this might be too much of a change...

File: py/picca/io.py
Comment (Waelthus) [reply py/picca/io.py:609]: I don't know if somebody uses it, but agree with Ignasi about the warning

File: py/picca/io.py
Comment (Waelthus) [reply py/picca/io.py:580]: I tend to prefer the f-string for variable names or very short operations and the format for longer operations, in this case I'd be on @julianbautista's side

Comment (iprafols) [reply py/picca/io.py:580]: ok, let's keep it with the f format then

File: bin/picca_deltas.py
Comment (iprafols) [reply bin/picca_deltas.py:489]: Yes, I think it would be too much of a change

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py:609]: added a warning

File: bin/picca_deltas.py
Comment (julianbautista) [reply bin/picca_deltas.py]: fitsio.read gives a recarray when reading a table, not an HDU. But I'm calling it hdu now.

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py]: In this part of the code, there cannot be two spectra for one single object/targetid. I think the sum was to convert an 2D array into a 1D array. 
I use np.where instead of a mask array so this sum() is not needed. I tested and it works.

**Review by iprafols — COMMENTED** [review]
Changes look fine

File: py/picca/io.py
Comment (Waelthus) [reply py/picca/io.py]: Ok, I only had this ever do the 2d to 1d conversion (it definitely does so with current data). Personally I like the mask array syntax more than np.where, but that's probably a question of taste.
If I'm not misreading the code having two spectra of the same target on the same tile in the same night would lead to coadding them in the old version, but would just pick the first of them in the new. Given that the outputs we're using are already coadded this should never happen. Might still be worth putting in a check and issuing a warning/raising an exception if there's ever more than one index in w_t.

I'm still against the "*1" as copy() is clearer to the reader (but I do not think the copying is even necessary in this case), i.e. a future maintainer might be tempted to remove a "*1" assuming it does nothing.

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py:489]: In this case I'd opt for @iprafols suggestion to add a set_order method. Also: rename the `pix` in the for loops to healpix for consistency

File: py/picca/io.py
Comment (julianbautista) [reply py/picca/io.py]: I've added the warning and added .copy() for each array.

File: bin/picca_deltas.py
Comment (iprafols) [reply bin/picca_deltas.py]: but does it not have a header? If it is a recarray you can rename it to something else

**Comment by julianbautista** [comment]
sorry I don't think this is that important. Feel free to change it in the
future.

**Comment by julianbautista** [comment]
I tested this branch and I obtain the same results as master with
- eBOSS DR16 data on Stripe82 
- eBOSS quickquasar Saclay mocks (by Julianna and Thomas)

I'm still waiting for NERSC to come back so I can test the deltas in P1D format by Michael.

**Review by iprafols — COMMENTED** [review]
changes look fine

File: py/picca/io.py
Comment (iprafols) [reply py/picca/io.py]: This is addressed with commit d464ea8

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py:528]: Ok, did the changes, it affects `data.py`, `picca_deltas.py` and `picca_plotSpec.py`. Should I commit it in here, or should we just leave this as is and I open a PR?

File: py/picca/io.py
Comment (Waelthus) [reply py/picca/io.py:157]: ok, the current approach seems to work on with a drq-style desi mock catalog and the london mocks, so I'd consider this solved.

File: py/picca/io.py
Comment (Waelthus) [inline py/picca/io.py:630]: wouldn't it be better to use `warnings.warn("this method will be deprecated.", DeprecationWarning)`. No show stopper, but might be good practice for the future

File: bin/picca_deltas.py
Comment (julianbautista) [reply bin/picca_deltas.py:528]: I think you can commit it here. Thanks!

**Review by Waelthus — APPROVED** [review]
Ok, given that tests run and all remaining changes are minor cosmetics, I'd say it's good to go. Probably issue a warning that output file structure changed for `mode==desi` due to a change in nside from hardcoded to automatically estimated, but everything else can easily be done later in seperate PRs.

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py]: Done!

File: bin/picca_deltas.py
Comment (julianbautista) [inline bin/picca_deltas.py:505]: Thanks for these changes. This can be even simpler thanks to the dictionary-like behaviour of Tables. You can do:
```
mask['log_wave_min'] = np.log10(mask['wave_min'])
mask['log_wave_max'] = np.log10(mask['wave_max'])
```

**Comment by Waelthus** [comment]
Oh cool, thanks! Didn't even know that this works ;) feel free to do the change

On October 1, 2020 2:12:20 PM GMT+02:00, Julian Bautista <notifications@github.com> wrote:
>@julianbautista commented on this pull request.
>
>
>
>> -                    np.log10(mask['wave_min'])[w],
>-                    np.log10(mask['wave_max'])[w]
>-                ]).T
>-            else:
>-                mask_rest_frame = np.array([])
>-            if 'RF_DLA' in mask['frame']:
>-                w = mask['frame'] == 'RF_DLA'
>-                mask_rest_frame_dla = np.array([
>-                    np.log10(mask['wave_min'])[w],
>-                    np.log10(mask['wave_max'])[w]
>-                ]).T
>-            else:
>-                mask_rest_frame_dla = None
>+            col_log_min=Column(np.log10(mask['wave_min']),name='log_wave_min')
>+            col_log_max=Column(np.log10(mask['wave_max']),name='log_wave_max')
>+            mask.add_columns([col_log_min,col_log_max])
>
>Thanks for these changes. This can be even simpler thanks to the dictionary-like behaviour of Tables. You can do:
>```
>mask['log_wave_min'] = np.log10(mask['wave_min'])
>mask['log_wave_max'] = np.log10(mask['wave_max'])
>```
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/igmhub/picca/pull/745#pullrequestreview-500281047
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

File: bin/picca_deltas.py
Comment (Waelthus) [reply bin/picca_deltas.py]: indeed, this is even better, done!

**Comment by julianbautista** [comment]
Can we merge this now ?

**Review by iprafols — APPROVED** [review]

**Pull request merged**
